elastic/apm-server
# Add support for Composite spans in the intake API

## Motivation/summary

Support a composite object in the span model and in the intake API. The composite object holds metrics on a group of spans represented by a single one.

The composite object has the following format

```
{
   "name": "SELECT FROM product_types WHERE user_id=?",
   "timestamp": 1613575086000000,
   "duration": 3781,
   "composite": {
       	"count": 10,
        "sum": 3592,
        "compression_strategy": "exact_match"
    }
}
```

## Checklist


## How to test these changes

To test these changes specifically, not including the changes to span metrics calculations:

- `make`
- `docker compose up -d`
- Add the ES credentials to `apm-server.yml` (`username: "admin"`, `password: "changeme"`)
- `./apm-server -c apm-server.yml -e -d "*"`
- `curl -H "Content-type: application/x-ndjson" --data-binary @testdata/intake-v2/spans.ndjson http://localhost:8200/intake/v2/events`
- In Kibana Dev Tools, run the following query and verify the span document has the expected `composite` nested object.
```
GET /apm-8.0.0-span/_search
{
  "query": {
    "match": {
      "trace.id": "edcbaf0123456789abcdef9876543210"
    }
  }
}
```
Expected nested document:
```
            "composite" : {
              "count" : 10,
              "sum" : {
                "us" : 359298
              },
              "compression_strategy" : "exact_match"
            }
```

This can also be tested along with https://github.com/elastic/apm-server/issues/5595. 


## Related issues

Related #5485 
Related #5595
